File: mobile/modules/unity-ads/android/src/main/java/expo/modules/unityads/UnityAdsModule.kt
package expo.modules.unityads

import android.content.pm.ApplicationInfo
import android.util.Log
import expo.modules.kotlin.Promise
import expo.modules.kotlin.functions.Queues
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.unity3d.ads.IUnityAdsInitializationListener
import com.unity3d.ads.UnityAds

class UnityAdsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoUnityAds")

    AsyncFunction("initialize") { gameId: String, testMode: Boolean, promise: Promise ->
      if (isInitialized && initializedGameId == gameId) {
        Log.i(
          TAG,
          "initialize skipped (already initialized) gameId=$gameId " +
            "unityIsInitialized=${UnityAds.isInitialized} testModeWas=$lastTestMode",
        )
        promise.resolve(true)
        return@AsyncFunction
      }

      val reactContext = requireNotNull(appContext.reactContext)
      val debuggable =
        (reactContext.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
      // Debug APKs embed a production JS bundle (__DEV__=false). Always enable Unity
      // test mode when the Android package is debuggable so banners actually fill.
      val effectiveTestMode = testMode || debuggable
      lastTestMode = effectiveTestMode

      Log.i(
        TAG,
        "initialization starting gameId=$gameId testMode=$effectiveTestMode " +
          "(jsTestMode=$testMode debuggable=$debuggable alreadyInitialized=${UnityAds.isInitialized})",
      )
      UnityAds.debugMode = debuggable

      UnityAds.initialize(
        reactContext.applicationContext,
        gameId,
        effectiveTestMode,
        object : IUnityAdsInitializationListener {
          override fun onInitializationComplete() {
            initializedGameId = gameId
            isInitialized = true
            Log.i(
              TAG,
              "initialization success gameId=$gameId testMode=$effectiveTestMode " +
                "unityIsInitialized=${UnityAds.isInitialized}",
            )
            promise.resolve(true)
          }

          override fun onInitializationFailed(
            error: UnityAds.UnityAdsInitializationError,
            message: String,
          ) {
            isInitialized = false
            Log.e(
              TAG,
              "initialization failure gameId=$gameId testMode=$effectiveTestMode " +
                "error=$error message=$message unityIsInitialized=${UnityAds.isInitialized}",
            )
            promise.reject(
              "ERR_UNITY_ADS_INIT",
              "Unity Ads failed to initialize: $error - $message",
              null,
            )
          }
        },
      )
    }.runOnQueue(Queues.MAIN)

    View(UnityAdsView::class) {
      Events("onAdLoaded", "onAdFailedToLoad")

      Prop("placementId") { view: UnityAdsView, placementId: String? ->
        view.setPlacementId(placementId)
      }

      OnViewDestroys { view: UnityAdsView ->
        Log.i(TAG, "component unmount (native OnViewDestroys)")
        view.destroyBanner("OnViewDestroys")
      }
    }
  }

  companion object {
    private const val TAG = "HaserliUnityAds"
    private var initializedGameId: String? = null
    private var isInitialized = false
    private var lastTestMode: Boolean? = null
  }
}

File: mobile/modules/unity-ads/ios/UnityAdsModule.swift
import ExpoModulesCore
import UnityAds

private final class UnityAdsInitDelegate: NSObject, UnityAdsInitializationDelegate {
  private let gameId: String
  private let promise: Promise
  private let onFinished: (UnityAdsInitDelegate) -> Void

  init(
    gameId: String,
    promise: Promise,
    onFinished: @escaping (UnityAdsInitDelegate) -> Void
  ) {
    self.gameId = gameId
    self.promise = promise
    self.onFinished = onFinished
  }

  func initializationComplete() {
    UnityAdsModule.markInitialized(gameId: gameId)
    promise.resolve(true)
    onFinished(self)
  }

  func initializationFailed(
    _ error: UnityAdsInitializationError,
    withMessage message: String
  ) {
    UnityAdsModule.markInitializationFailed()
    promise.reject(
      "ERR_UNITY_ADS_INIT",
      "Unity Ads failed to initialize: \(message)",
      nil
    )
    onFinished(self)
  }
}

public class UnityAdsModule: Module {
  // Retain init delegates until the callback fires.
  private static var pendingInitDelegates = [UnityAdsInitDelegate]()
  private static var initializedGameId: String?
  private static var isInitialized = false

  static func markInitialized(gameId: String) {
    initializedGameId = gameId
    isInitialized = true
  }

  static func markInitializationFailed() {
    isInitialized = false
  }

  public func definition() -> ModuleDefinition {
    Name("ExpoUnityAds")

    AsyncFunction("initialize") { (gameId: String, testMode: Bool, promise: Promise) in
      if Self.isInitialized, Self.initializedGameId == gameId {
        promise.resolve(true)
        return
      }

      let delegate = UnityAdsInitDelegate(gameId: gameId, promise: promise) { finished in
        Self.pendingInitDelegates.removeAll { $0 === finished }
      }
      Self.pendingInitDelegates.append(delegate)

      UnityAds.initialize(
        gameId,
        testMode: testMode,
        initializationDelegate: delegate
      )
    }
    .runOnQueue(.main)

    View(UnityAdsView.self) {
      Events("onAdLoaded", "onAdFailedToLoad")

      Prop("placementId") { (view: UnityAdsView, placementId: String?) in
        view.setPlacementId(placementId)
      }
    }
  }
}
